### ISSUE TITLE
kairos-agent cannot upgrade recovery image using custom image source - Error sanitizing the upgradepec: undefined upgrade source 

### ISSUE THREAD

**Issue by vadimzharov** [kairos-io/kairos#1800]
**Kairos version:**

kairos-agent version v2.2.8

PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
KAIROS_NAME="kairos-vadim-ubuntu"
KAIROS_VERSION="v1.0.0"
KAIROS_ID="kairos"
KAIROS_ID_LIKE="kairos-vadim-ubuntu"
KAIROS_VERSION_ID="v1.0.0"
KAIROS_PRETTY_NAME="kairos-vadim-ubuntu v1.0.0"
KAIROS_BUG_REPORT_URL="https://github.com/vadimzharov/k8s-kairos/issues"
KAIROS_HOME_URL="https://github.com/vadimzharov/k8s-kairos
KAIROS_IMAGE_REPO="quay.io/vadimzharov/kairos"
KAIROS_IMAGE_LABEL="latest"
KAIROS_GITHUB_REPO="vadimzharov/k8s-kairos"
KAIROS_VARIANT="core"
KAIROS_FLAVOR="ubuntu"

**CPU architecture, OS, and Version:**

Linux k8s-master 6.2.0-31-generic #31~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Wed Aug 16 13:45:26 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

**Describe the bug**

I'm using custom built Kairos image with kairos-agent version v2.2.8 and I'm trying to upgrade Kairos.
I successfully executed 
`kairos-agent upgrade --source oci:quay.io/vadimzharov/kairos:v1.28.1`
and my image has been successfully upgraded.
However, when I'm trying to upgrade the recovery image by executing the command:
`kairos-agent upgrade --recovery --source oci:quay.io/vadimzharov/kairos:v1.28.1`
I'm getting the following error:
undefined upgrade source

and DEBUG message in the output (see logs below):
WARN[2023-09-08T03:43:00Z] Error sanitizing the upgradepec: undefined upgrade source 

**To Reproduce**

execute 
`kairos-agent upgrade --recovery --source <any image>`
and the agent is ignoring the custom source for the image (all of them - oci, file,dir).

**Expected behavior**

kairos-agent is able to upgrade the recovery image using the provided source image.

**Logs**

```
root@k8s-master:/home/kairos# kairos-agent upgrade --recovery --source oci:quay.io/vadimzharov/kairos:v1.28.1
warning: skipping /oem/10_user_custom_mounts.yaml because it has no valid header
warning: skipping /oem/userdata because it has no valid header
INFO[2023-09-08T03:42:59Z] kairos-agent version v2.2.8                  
DEBU[2023-09-08T03:42:59Z] Loaded config: &config.Config{
  Install: &config.Install{
    Auto: true,
    Reboot: true,
    NoFormat: false,
    Device: "/dev/vda",
    Poweroff: false,
    GrubOptions: map[string]string(nil), // p0
    Bundles: nil,
    Encrypt: nil,
    SkipEncryptCopyPlugins: false,
    Env: nil,
    Image: "",
    EphemeralMounts: nil,
    BindMounts: []string{
      "/etc/kubernetes/",
      "/var/lib/etcd/",
    },
  },
  Config: collector.Config{
    "debug": true,
    "growpart": collector.Config{
      "devices": []interface {}{
        "/",
      },
    },
    "hostname": "k8s-master",
    "install": collector.Config{
      "auto": true,
      "bind_mounts": []interface {}{
        "/etc/kubernetes/",
        "/var/lib/etcd/",
      },
      "device": "/dev/vda",
      "firmware": "bios",
      "force": true,
      "grub-entry-name": "Kairos",
      "no-format": false,
      "part-table": "gpt",
      "partitions": collector.Config{
        "oem": collector.Config{
          "fs": "ext4",
          "size": 60,
        },
        "recovery": collector.Config{
          "fs": "ext4",
          "size": 4096,
        },
      },
      "poweroff": false,
      "reboot": true,
    },
    "reset": collector.Config{
      "grub-entry-name": "Kairos",
      "poweroff": true,
      "reboot": true,
      "reset-oem": false,
      "reset-persistent": true,
    },
    "stages": collector.Config{
      "config": []interface {}{
        collector.Config{
          "name": "Create kubeadm-config",
          "write_files": []interface {}{
            collector.Config{
              "content": "XXXX",
              "encoding": "b64",
              "owner": "root",
              "path": "/etc/kubernetes/kubeadm-config.yaml",
              "permissions": "0644",
            },
          },
        },
      },
      "network": []interface {}{
        collector.Config{
          "authorized_keys": collector.Config{
            "kairos": []interface {}{
              "ssh-rsa XXXX",
            },
          },
          "name": "Setup users",
        },
      },
    },
    "strict": true,
    "upgrade": collector.Config{
      "grub-entry-name": "Kairos",
      "poweroff": false,
      "reboot": true,
      "recovery": false,
    },
    "users": []interface {}{
      collector.Config{
        "lock_passwd": true,
        "name": "kairos",
        "passwd": "kairos",
        "ssh_authorized_keys": []interface {}{
          "ssh-rsa XXXX",
        },
      },
    },
    "write_files": []interface {}{
      collector.Config{
        "content": "XXX",
        "encoding": "b64",
        "owner": "root",
        "path": "/etc/kubernetes/kubeadm-config.yaml",
        "permissions": "0644",
      },
    },
  },
  ConfigURL: "",
  Options: p0,
  FailOnBundleErrors: false,
  Bundles: nil,
  GrubOptions: p0,
  Env: nil,
  Debug: true,
  Strict: true,
  CloudInitPaths: nil,
  EjectCD: false,
  Logger: &logrus.Logger{ // p1
    Out: &os.File{},
    Hooks: logrus.LevelHooks{},
    Formatter: &logrus.TextFormatter{
      ForceColors: true,
      DisableColors: false,
      ForceQuote: false,
      DisableQuote: false,
      EnvironmentOverrideColors: false,
      DisableTimestamp: false,
      FullTimestamp: true,
      TimestampFormat: "",
      DisableSorting: false,
      SortingFunc: ,
      DisableLevelTruncation: false,
      PadLevelText: false,
      QuoteEmptyFields: false,
      FieldMap: logrus.FieldMap(nil),
      CallerPrettyfier: ,
    },
    ReportCaller: false,
    Level: 5,
    ExitFunc: os.Exit,
    BufferPool: nil,
  },
  Fs: &vfs.osfs{}, // p2
  Mounter: &mount.Mounter{},
  Runner: &v1.RealRunner{ // p3
    Logger: p1,
  },
  Syscall: &v1.RealSyscall{},
  CloudInitRunner: &cloudinit.YipCloudInitRunner{},
  ImageExtractor: v1.OCIImageExtractor{},
  Client: &http.Client{},
  Platform: &v1.Platform{
    OS: "linux",
    Arch: "x86_64",
    GolangArch: "amd64",
  },
  Cosign: false,
  Verify: false,
  CosignPubKey: "",
  Arch: "x86_64",
  SquashFsCompressionConfig: []string{},
  SquashFsNoCompression: true,
} 
INFO[2023-09-08T03:43:00Z] Setting image size to 3608Mb                 
WARN[2023-09-08T03:43:00Z] Error sanitizing the upgradepec: undefined upgrade source 
DEBU[2023-09-08T03:43:00Z] Loaded upgrade spec: &v1.UpgradeSpec{
  RecoveryUpgrade: false,
  Active: v1.Image{
    File: "/run/initramfs/cos-state/cOS/transition.img",
    Label: "COS_ACTIVE",
    Size: 3072,
    FS: "ext2",
    Source: &v1.ImageSource{},
    MountPoint: "/run/cos/transition",
    LoopDevice: "",
  },
  Recovery: v1.Image{
    File: "/run/cos/recovery/cOS/transition.img",
    Label: "COS_SYSTEM",
    Size: 3608,
    FS: "ext2",
    Source: &v1.ImageSource{},
    MountPoint: "/run/cos/transition",
    LoopDevice: "",
  },
  GrubDefEntry: "Kairos",
  Reboot: true,
  PowerOff: false,
  ExtraDirsRootfs: nil,
  Passive: v1.Image{
    File: "/run/initramfs/cos-state/cOS/passive.img",
    Label: "COS_PASSIVE",
    Size: 3072,
    FS: "ext2",
    Source: &v1.ImageSource{},
    MountPoint: "",
    LoopDevice: "",
  },
  Partitions: v1.ElementalPartitions{
    BIOS: nil,
    EFI: nil,
    OEM: &v1.Partition{
      Name: "vda2",
      FilesystemLabel: "COS_OEM",
      Size: 60,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/oem",
      Path: "/dev/vda2",
      Disk: "/dev/vda",
    },
    Recovery: &v1.Partition{
      Name: "vda3",
      FilesystemLabel: "COS_RECOVERY",
      Size: 4096,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/run/cos/recovery",
      Path: "/dev/vda3",
      Disk: "/dev/vda",
    },
    State: &v1.Partition{
      Name: "vda4",
      FilesystemLabel: "COS_STATE",
      Size: 10063,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/run/initramfs/cos-state",
      Path: "/dev/vda4",
      Disk: "/dev/vda",
    },
    Persistent: &v1.Partition{
      Name: "vda5",
      FilesystemLabel: "COS_PERSISTENT",
      Size: 11378,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/run/cos/persistent",
      Path: "/dev/vda5",
      Disk: "/dev/vda",
    },
  },
  State: &v1.InstallState{
    Date: "2023-09-08T03:28:31Z",
    Partitions: map[string]*v1.PartitionState{
      "oem": &v1.PartitionState{
        FSLabel: "COS_OEM",
        Images: map[string]*v1.ImageState(nil), // p0
      },
      "persistent": &v1.PartitionState{
        FSLabel: "COS_PERSISTENT",
        Images: p0,
      },
      "recovery": &v1.PartitionState{
        FSLabel: "COS_RECOVERY",
        Images: map[string]*v1.ImageState{
          "recovery": &v1.ImageState{
            Source: &v1.ImageSource{},
            SourceMetadata: nil,
            Label: "COS_SYSTEM",
            FS: "ext2",
          },
        },
      },
      "state": &v1.PartitionState{
        FSLabel: "COS_STATE",
        Images: map[string]*v1.ImageState{
          "active": &v1.ImageState{
            Source: &v1.ImageSource{},
            SourceMetadata: nil,
            Label: "COS_ACTIVE",
            FS: "ext2",
          },
          "passive": &v1.ImageState{
            Source: &v1.ImageSource{},
            SourceMetadata: nil,
            Label: "COS_ACTIVE",
            FS: "ext2",
          },
        },
      },
    },
  },
} 
undefined upgrade source
```

**Additional context**

**Comment by Itxaka** [comment]
umm, I just upgraded to your image, rebooted into it, all ok. Then upgraded the recovery and had no issues...


```
root@localhost:~# kairos-agent version --long
{Version:v2.2.8 GitCommit:none GoVersion:go1.20.2}
root@localhost:~#  kairos-agent upgrade --recovery --source oci:quay.io/vadimzharov/kairos:v1.28.1
warning: skipping /oem/file (extension).
WARNING: jsonschema: '' does not validate with file:///root/schema.json#/required: missing properties: 'users'
INFO[2023-09-12T10:12:47Z] kairos-agent version v2.2.8                  
INFO[2023-09-12T10:12:48Z] Setting image size to 3608Mb                 
INFO[2023-09-12T10:12:48Z] Applying 'before-upgrade' hook               
INFO[2023-09-12T10:12:48Z] Running before-upgrade hook                  
INFO[2023-09-12T10:12:48Z] Running stage: before-upgrade.before         
INFO[2023-09-12T10:12:48Z] Done executing stage 'before-upgrade.before' 
INFO[2023-09-12T10:12:48Z] Running stage: before-upgrade                
INFO[2023-09-12T10:12:48Z] Done executing stage 'before-upgrade'        
INFO[2023-09-12T10:12:48Z] Running stage: before-upgrade.after          
INFO[2023-09-12T10:12:48Z] Done executing stage 'before-upgrade.after'  
INFO[2023-09-12T10:12:48Z] Running stage: before-upgrade.before         
INFO[2023-09-12T10:12:48Z] Done executing stage 'before-upgrade.before' 
INFO[2023-09-12T10:12:48Z] Running stage: before-upgrade                
INFO[2023-09-12T10:12:48Z] Done executing stage 'before-upgrade'        
INFO[2023-09-12T10:12:48Z] Running stage: before-upgrade.after          
INFO[2023-09-12T10:12:48Z] Done executing stage 'before-upgrade.after'  
INFO[2023-09-12T10:12:48Z] deploying image quay.io/vadimzharov/kairos:v1.28.1 to /run/cos/recovery/cOS/transition.img 
INFO[2023-09-12T10:12:48Z] Creating file system image /run/cos/recovery/cOS/transition.img with size 3608Mb 
INFO[2023-09-12T10:12:48Z] Copying quay.io/vadimzharov/kairos:v1.28.1 source to /run/cos/transition 
INFO[2023-09-12T10:13:49Z] Finished copying quay.io/vadimzharov/kairos:v1.28.1 into /run/cos/transition 
INFO[2023-09-12T10:13:49Z] Applying 'after-upgrade-chroot' hook         
INFO[2023-09-12T10:13:49Z] Running after-upgrade-chroot hook            
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade-chroot.before   
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade-chroot.before' 
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade-chroot          
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade-chroot'  
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade-chroot.after    
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade-chroot.after' 
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade-chroot.before   
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade-chroot.before' 
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade-chroot          
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade-chroot'  
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade-chroot.after    
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade-chroot.after' 
INFO[2023-09-12T10:13:49Z] Moving /run/cos/recovery/cOS/transition.img to /run/cos/recovery/cOS/recovery.img 
INFO[2023-09-12T10:13:49Z] Finished moving /run/cos/recovery/cOS/transition.img to /run/cos/recovery/cOS/recovery.img 
INFO[2023-09-12T10:13:49Z] Applying 'after-upgrade' hook                
INFO[2023-09-12T10:13:49Z] Running after-upgrade hook                   
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade.before          
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade.before'  
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade                 
INFO[2023-09-12T10:13:49Z] Processing stage step 'Mount state'. ( commands: 1, files: 0, ... ) 
INFO[2023-09-12T10:13:49Z] Command output:                              
WARN[2023-09-12T10:13:49Z] (conditional) Skip 'Skipping stage (if statement error: failed to run [ $(kairos-agent state get oem.found) == 'true' ]: exit status 2)' stage name: Update plugins 
INFO[2023-09-12T10:13:49Z] Processing stage step 'Set upgrade sentinel'. ( commands: 1, files: 0, ... ) 
INFO[2023-09-12T10:13:49Z] Command output:                              
WARN[2023-09-12T10:13:49Z] (conditional) Skip 'Skipping stage (if statement error: failed to run ! grep -q "grub_boot_assessment" /tmp/mnt/STATE/grubcustom
: exit status 1)' stage name: Hook boot assessment grub configuration 
INFO[2023-09-12T10:13:49Z] Processing stage step 'Grub branding'. ( commands: 1, files: 0, ... ) 
INFO[2023-09-12T10:13:49Z] Command output: '/etc/kairos/branding/grubmenu.cfg' -> '/tmp/mnt/STATE/grubmenu' 
INFO[2023-09-12T10:13:49Z] Processing stage step 'Add boot assessment grub configuration'. ( commands: 0, files: 1, ... ) 
INFO[2023-09-12T10:13:49Z] Processing stage step 'umount state'. ( commands: 1, files: 0, ... ) 
INFO[2023-09-12T10:13:49Z] Command output:                              
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade'         
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade.after           
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade.after'   
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade.before          
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade.before'  
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade                 
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade'         
INFO[2023-09-12T10:13:49Z] Running stage: after-upgrade.after           
INFO[2023-09-12T10:13:49Z] Done executing stage 'after-upgrade.after'   
INFO[2023-09-12T10:13:49Z] Upgrade completed                            
 INFO  Shutdown node
root@localhost:~# cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
KAIROS_NAME="kairos-vadim-ubuntu"
KAIROS_VERSION="v1.0.0"
KAIROS_ID="kairos"
KAIROS_ID_LIKE="kairos-vadim-ubuntu"
KAIROS_VERSION_ID="v1.0.0"
KAIROS_PRETTY_NAME="kairos-vadim-ubuntu v1.0.0"
KAIROS_BUG_REPORT_URL="https://github.com/vadimzharov/k8s-rpi-cluster/issues"
KAIROS_HOME_URL="https://github.com/vadimzharov/k8s-rpi-cluster"
KAIROS_IMAGE_REPO="quay.io/vadimzharov/core-ubuntu"
KAIROS_IMAGE_LABEL="latest"
KAIROS_GITHUB_REPO="vadimzharov/k8s-rpi-cluster"
KAIROS_VARIANT="core"
KAIROS_FLAVOR="ubuntu"
root@localhost:~# 

```

**Comment by Itxaka** [comment]
You should not get that issue anymore, that its tested on master and doesnt seem to fail. Sorry no idea what else it could be.

Notice that unless you are adding extra stuff that you need on recovery, you can upgrade the recovery to plain kairos without needing to use your own image, as recovery is meant only to reset the system or salvage stuff with manual intervention, its not a required step for upgrade as recovery should be the last resource to recover a system.

Also if you generate an iso with Auroraboot or osbuilder based of your image, your recovery will already be based off on your image, so no need to upgrade it.

**Comment by vadimzharov** [comment]
Thank you!
This is the case I'm trying to test - have a long-live k8s cluster which I can reset if needed without reinstalling the OS/reprovision VMs. The steps I'm following:
1. Build my own Kairos image with k8s binaries (i.e. 1.28.0) using custom image + AuroraBoot for ISO
2. Install k8s using this image
3. Build a new image with updated k8s binaries (i.e 1.28.1)
4. Upgrade my k8s cluster using the updated image
......
5. Reset my k8s cluster without the requirement to reprovision OS.

So right now I cannot do this after I upgrade the OS using the new image - I'm getting 
```
kairos-agent reset --unattended
failed initializing spec: reset can only be called from the recovery system
```
And with this bug - I need to add a step before resetting the state - go back to the recovery. And since my recovery image is not upgraded - I need to:
1. Go back to the recovery image
2. Reset the OS using `kairos-agent reset`
3. Upgrade the OS using `kairos-agent upgrade`
4. Install k8s cluster.
This just adds one more step, which is not a big deal, but without it the process is more lightweight.

Thank you, I will try to test the agent from the master branch (or when it will be released).

**Comment by jimmykarily** [comment]
@vadimzharov 2.4.0 is out. Please give a try when you find time and let us know if the issue is fixed. Thanks!

**Comment by vadimzharov** [comment]
Hi @jimmykarily ! I rebuilt my images but the issue still exists.
The Dockerfile I used: [https://github.com/vadimzharov/k8s-talos/blob/main/Dockerfile.00](https://github.com/vadimzharov/k8s-talos/blob/main/Dockerfile.00)
```
root@k8s-master:/home/kairos# kairos-agent --version
kairos-agent version v2.2.12
```

But the issue is still there:

```
root@k8s-master:/home/kairos# kairos-agent upgrade --recovery --source oci:quay.io/vadimzharov/kairos:v1.28.1
warning: skipping /oem/10_user_custom_mounts.yaml because it has no valid header
INFO[2023-09-18T15:57:30Z] kairos-agent version v2.2.12                 
DEBU[2023-09-18T15:57:30Z] Loaded config: &config.Config{
  Install: &config.Install{
    Auto: true,
    Reboot: true,
    NoFormat: false,
    Device: "/dev/vda",
    Poweroff: false,
    GrubOptions: map[string]string(nil), // p0
    Bundles: nil,
    Encrypt: nil,
    SkipEncryptCopyPlugins: false,
    Env: nil,
    Image: "",
    EphemeralMounts: nil,
    BindMounts: []string{
      "/etc/kubernetes/",
      "/var/lib/etcd/",
      "/var/lib/kubelet/",
    },
  },
  Config: collector.Config{
    "debug": true,
    "growpart": collector.Config{
      "devices": []interface {}{
        "/",
      },
    },
    "hostname": "k8s-master",
    "install": collector.Config{
      "auto": true,
      "bind_mounts": []interface {}{
        "/etc/kubernetes/",
        "/var/lib/etcd/",
        "/var/lib/kubelet/",
      },
      "device": "/dev/vda",
      "firmware": "bios",
      "force": true,
      "grub-entry-name": "Kairos",
      "no-format": false,
      "part-table": "gpt",
      "partitions": collector.Config{
        "oem": collector.Config{
          "fs": "ext4",
          "size": 60,
        },
        "recovery": collector.Config{
          "fs": "ext4",
          "size": 4096,
        },
      },
      "poweroff": false,
      "reboot": true,
    },
    "reset": collector.Config{
      "grub-entry-name": "Kairos",
      "poweroff": true,
      "reboot": true,
      "reset-oem": false,
      "reset-persistent": true,
    },
    "strict": true,
    "upgrade": collector.Config{
      "grub-entry-name": "Kairos",
      "poweroff": false,
      "reboot": true,
      "recovery": false,
    },
    "users": []interface {}{
      collector.Config{
        "lock_passwd": true,
        "name": "kairos",
        "passwd": "kairos",
        "ssh_authorized_keys": []interface {}{
          "XXX",
        },
      },
    },
    "write_files": []interface {}{
      collector.Config{
        "content": "XXX",
        "encoding": "b64",
        "owner": "root",
        "path": "/etc/kubernetes/kubeadm-config.yaml",
        "permissions": "0644",
      },
    },
  },
  ConfigURL: "",
  Options: p0,
  FailOnBundleErrors: false,
  Bundles: nil,
  GrubOptions: p0,
  Env: nil,
  Debug: true,
  Strict: true,
  CloudInitPaths: nil,
  EjectCD: false,
  Logger: &logrus.Logger{ // p1
    Out: &os.File{},
    Hooks: logrus.LevelHooks{},
    Formatter: &logrus.TextFormatter{
      ForceColors: true,
      DisableColors: false,
      ForceQuote: false,
      DisableQuote: false,
      EnvironmentOverrideColors: false,
      DisableTimestamp: false,
      FullTimestamp: true,
      TimestampFormat: "",
      DisableSorting: false,
      SortingFunc: ,
      DisableLevelTruncation: false,
      PadLevelText: false,
      QuoteEmptyFields: false,
      FieldMap: logrus.FieldMap(nil),
      CallerPrettyfier: ,
    },
    ReportCaller: false,
    Level: 5,
    ExitFunc: os.Exit,
    BufferPool: nil,
  },
  Fs: &vfs.osfs{}, // p2
  Mounter: &mount.Mounter{},
  Runner: &v1.RealRunner{ // p3
    Logger: p1,
  },
  Syscall: &v1.RealSyscall{},
  CloudInitRunner: &cloudinit.YipCloudInitRunner{},
  ImageExtractor: v1.OCIImageExtractor{},
  Client: &http.Client{},
  Platform: &v1.Platform{
    OS: "linux",
    Arch: "x86_64",
    GolangArch: "amd64",
  },
  Cosign: false,
  Verify: false,
  CosignPubKey: "",
  Arch: "x86_64",
  SquashFsCompressionConfig: []string{},
  SquashFsNoCompression: true,
} 
INFO[2023-09-18T15:57:31Z] Setting image size to 3634Mb                 
WARN[2023-09-18T15:57:31Z] Error sanitizing the upgradepec: undefined upgrade source 
DEBU[2023-09-18T15:57:31Z] Loaded upgrade spec: &v1.UpgradeSpec{
  RecoveryUpgrade: false,
  Active: v1.Image{
    File: "/run/initramfs/cos-state/cOS/transition.img",
    Label: "COS_ACTIVE",
    Size: 3072,
    FS: "ext2",
    Source: &v1.ImageSource{},
    MountPoint: "/run/cos/transition",
    LoopDevice: "",
  },
  Recovery: v1.Image{
    File: "/run/cos/recovery/cOS/transition.img",
    Label: "COS_SYSTEM",
    Size: 3634,
    FS: "ext2",
    Source: &v1.ImageSource{},
    MountPoint: "/run/cos/transition",
    LoopDevice: "",
  },
  GrubDefEntry: "Kairos",
  Reboot: true,
  PowerOff: false,
  ExtraDirsRootfs: nil,
  Passive: v1.Image{
    File: "/run/initramfs/cos-state/cOS/passive.img",
    Label: "COS_PASSIVE",
    Size: 3072,
    FS: "ext2",
    Source: &v1.ImageSource{},
    MountPoint: "",
    LoopDevice: "",
  },
  Partitions: v1.ElementalPartitions{
    BIOS: nil,
    EFI: nil,
    OEM: &v1.Partition{
      Name: "vda2",
      FilesystemLabel: "COS_OEM",
      Size: 60,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/oem",
      Path: "/dev/vda2",
      Disk: "/dev/vda",
    },
    Recovery: &v1.Partition{
      Name: "vda3",
      FilesystemLabel: "COS_RECOVERY",
      Size: 4096,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/run/cos/recovery",
      Path: "/dev/vda3",
      Disk: "/dev/vda",
    },
    State: &v1.Partition{
      Name: "vda4",
      FilesystemLabel: "COS_STATE",
      Size: 10081,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/run/initramfs/cos-state",
      Path: "/dev/vda4",
      Disk: "/dev/vda",
    },
    Persistent: &v1.Partition{
      Name: "vda5",
      FilesystemLabel: "COS_PERSISTENT",
      Size: 11360,
      FS: "ext4",
      Flags: nil,
      MountPoint: "/run/cos/persistent",
      Path: "/dev/vda5",
      Disk: "/dev/vda",
    },
  },
  State: &v1.InstallState{
    Date: "2023-09-18T15:56:00Z",
    Partitions: map[string]*v1.PartitionState{
      "oem": &v1.PartitionState{
        FSLabel: "COS_OEM",
        Images: map[string]*v1.ImageState(nil), // p0
      },
      "persistent": &v1.PartitionState{
        FSLabel: "COS_PERSISTENT",
        Images: p0,
      },
      "recovery": &v1.PartitionState{
        FSLabel: "COS_RECOVERY",
        Images: map[string]*v1.ImageState{
          "recovery": &v1.ImageState{
            Source: &v1.ImageSource{},
            SourceMetadata: nil,
            Label: "COS_SYSTEM",
            FS: "ext2",
          },
        },
      },
      "state": &v1.PartitionState{
        FSLabel: "COS_STATE",
        Images: map[string]*v1.ImageState{
          "active": &v1.ImageState{
            Source: &v1.ImageSource{},
            SourceMetadata: nil,
            Label: "COS_ACTIVE",
            FS: "ext2",
          },
          "passive": &v1.ImageState{
            Source: &v1.ImageSource{},
            SourceMetadata: nil,
            Label: "COS_ACTIVE",
            FS: "ext2",
          },
        },
      },
    },
  },
} 
undefined upgrade source
```

**Comment by vadimzharov** [comment]
This is the content and `ls` for all directories:
```
root@k8s-master:/home/kairos# cat /oem/10_user_custom_mounts.yaml
stages:
  rootfs:
  - dns: {}
    name: user_custom_mounts
    systemctl: {}
    environment:
      CUSTOM_BIND_MOUNTS: /etc/kubernetes/ /var/lib/etcd/ /var/lib/kubelet/
      CUSTOM_EPHEMERAL_MOUNTS: ""
    environment_file: /run/cos/custom-layout.env
    datasource: {}
    layout: {}
    git:
      auth: {}
root@k8s-master:/home/kairos# ls /etc/kubernetes/
admin.conf  controller-manager.conf  kubeadm-config.yaml  kubelet.conf  manifests  pki  scheduler.conf
root@k8s-master:/home/kairos# ls /var/lib/etcd/
member
root@k8s-master:/home/kairos# ls /var/lib/kubelet/
config.yaml  cpu_manager_state  device-plugins  kubeadm-flags.env  memory_manager_state  pki  plugins  plugins_registry  pod-resources  pods
root@k8s-master:/home/kairos# 
```

**Comment by jimmykarily** [comment]
Never mind, it's a bug. `--source` is not respected. I created an issue to fix this: https://github.com/kairos-io/kairos/issues/1837

**Comment by jimmykarily** [comment]
> Never mind, it's a bug. `--source` is not respected. I created an issue to fix this: #1837

Actually not (see the conversation on the issue). We still need to reproduce this one.

**Comment by vadimzharov** [comment]
@jimmykarily 
The steps I'm following are described here - https://github.com/vadimzharov/k8s-kairos-deploy
The image I built is available as quay.io/vadimzharov/kairos:v1.28.0, and I'm using AuroraBoot to create ISO and bootstrap a VM. 
Auroraboot command:
`docker run -v "$PWD"/build:/tmp/auroraboot -v /var/run/docker.sock:/var/run/docker.sock --rm -ti quay.io/kairos/auroraboot:v0.2.5 --set container_image=quay.io/vadimzharov/kairos:v1.28.0 --set "disable_http_server=true" --set "disable_netboot=true" --set "state_dir=/tmp/auroraboot" --debug
`
How can I help to debut this issue?

**Comment by jimmykarily** [comment]
In the first comment (description), are you sure the output you sent comes from this command?

```
root@k8s-master:/home/kairos# kairos-agent upgrade --recovery --source oci:quay.io/vadimzharov/kairos:v1.28.1
``` 

I'm asking because the output is "debug" output but I don't see the `--debug` flag in your command. Also, this:

```
DEBU[2023-09-18T15:57:31Z] Loaded upgrade spec: &v1.UpgradeSpec{
  RecoveryUpgrade: false,
```

shows that `--recovery` was also ignored, otherwise it would have been "true".

See what happens when I run the command inside a docker container with your image:

```
~ $ docker run -it --rm quay.io/vadimzharov/kairos:v1.28.0 /bin/bash
root@2de9e0643ef1:/# kairos-agent --debug upgrade --recovery --source oci:quay.io/vadimzharov/kairos:v1.28.1
...
WARN[2023-09-27T06:28:33Z] Error sanitizing the upgradepec: undefined recovery partition 
DEBU[2023-09-27T06:28:33Z] Loaded upgrade spec: &v1.UpgradeSpec{
  RecoveryUpgrade: true,
...
undefined recovery partition
```
(skipped some output with `...`)

but you can see that `RecoveryUpgrade` is set to `true` which means the flag was respected. For both `--recovery` and `--source` we use this viper "hack" :

https://github.com/kairos-io/kairos-agent/blob/677254e5701b8698c2befd2c7f51c8bef08efc0b/internal/agent/upgrade.go#L99-L100

to later read it here:
https://github.com/kairos-io/kairos-agent/blob/677254e5701b8698c2befd2c7f51c8bef08efc0b/pkg/config/spec.go#L267-L268

which uses viper as bit like a global variable which is bad in my opinion and we should refactor (I will try something as part of this: https://github.com/kairos-io/kairos-agent/pull/152) but I wonder if there are cases in which it doesn't work.

**Comment by jimmykarily** [comment]
Debug output could have been enabled with the `KAIROS_AGENT_DEBUG` environment variable. I still can't explain the ignored flags.

**Comment by jimmykarily** [comment]
> Debug output could have been enabled with the `KAIROS_AGENT_DEBUG` environment variable. I still can't explain the ignored flags.

I see you have it set in your config:

```
  Config: collector.Config{
    "debug": true,
```
ok this explains the debug output.

**Comment by Itxaka** [comment]
I think we fixed this on 2.4.3 and upwards up to 3.0.x

**Comment by Itxaka** [comment]
closing this, feel free to reopen if its still happening
